lib/nvector_eigen.h

```diff
@@ -0,0 +1,528 @@
+    auto z_vec = static_cast<VectorType*>(z->content);
+
+    for (unsigned int i=0; i<x->ops->nvgetlength(x); ++i)
+    {
+      auto val = std::abs((*x_vec)(i));
+      (val > c) ? (*z_vec)(i) = 1. : (*z_vec)(i) = 0.;
+    }
+  }
+
+
+
+  /// Sets the components of z to be the inverses of the components of x, with testing for zero values.
+  template<typename VectorType>
+  booleantype
+  N_VInvTest(N_Vector x, N_Vector z)
+  {
+    auto x_vec = static_cast<VectorType*>(x->content);
+    auto z_vec = static_cast<VectorType*>(z->content);
+
+    booleantype result = SUNTRUE;
+
+    for (unsigned int i=0; i<x->ops->nvgetlength(x); ++i)
+    {
+      auto val = (*x_vec)(i);
+      if (val != 0)
+        (*z_vec)(i) = 1./val;
+      else
+        result = SUNFALSE;
+    }
+    return result;
+  }
+
+
+
+  /// Computes the linear combination of nv vectors X.
+  template<typename VectorType>
+  int
+  N_VLinearCombination(int nv, realtype* c, N_Vector* X, N_Vector z)
+  {
+    // invalid number of vectors
+    if (nv < 1) return -1;
+
+    // should have called N_VScale in this case
+    if (nv == 1)
+    {
+      N_VScale<VectorType>(c[0], X[0], z);
+      return 0;
+    }
+
+    // should have called N_VLinearSum
+    if (nv == 2)
+    {
+      N_VLinearSum<VectorType>(c[0], X[0], c[1], X[1], z);
+      return 0;
+    }
+
+    // when nv > 2, start with linear sum and then keep adding to z
+    N_VLinearSum<VectorType>(c[0], X[0], c[1], X[1], z);
+    for (unsigned int i=2; i<nv; ++i)
+      N_VLinearSum<VectorType>(1, z, c[i], X[i], z);
+
+    return 0;
+  }
+
+
+
+  /// Scales and adds one vector to nv vectors.
+  template<typename VectorType>
+  int
+  N_VScaleAddMulti(int nv, realtype* c, N_Vector x, N_Vector* Y, N_Vector* Z)
+  {
+    // invalid number of vectors
+    if (nv < 1) return -1;
+
+    for (unsigned int i=0; i<nv; ++i)
+    {
+      N_VLinearSum<VectorType>(c[i], x, 1, Y[i], Z[i]);
+    }
+
+    return 0;
+  }
+
+
+
+  /// Computes the dot product of a vector with n_v other vectors.
+  template<typename VectorType>
+  int
+  N_VDotProdMulti(int nv, N_Vector x, N_Vector* Y, realtype* d)
+  {
+    // invalid number of vectors
+    if (nv < 1) return -1;
+
+    auto x_vec = static_cast<VectorType*>(x->content);
+
+    for (unsigned int i=0; i<nv; ++i)
+    {
+      auto y_vec = static_cast<VectorType*>(Y[i]->content);
+      d[i] = x_vec->dot(*y_vec);
+    }
+    return 0;
+  }
+
+
+
+  /// Scales each vector by a potentially different constant.
+  template<typename VectorType>
+  int
+  N_VScaleVectorArray(int nv, realtype* c, N_Vector* X, N_Vector* Z)
+  {
+    // invalid number of vectors
+    if (nv < 1) return -1;
+
+    for (unsigned int i=0; i<nv; ++i)
+    {
+      N_VScale<VectorType>(c[i], X[i], Z[i]);
+    }
+    return 0;
+  }
+
+}
+
+
+
+/// Function to create an N_Vector without allocating memory for the vector.
+template<typename VectorType>
+N_Vector
+create_empty_eigen_nvector()
+{
+  N_Vector v = N_VNewEmpty();
```

Comment:
OK, so here you create a generic object and that allocates the `ops` structure as well. You then set the elements of that structure to the right pointers. The questions to me are then:
* The documentation of `N_VNewEmpty` says "This allocates a new generic N_Vector object and initializes its content pointer and the function pointers in the operations structure to NULL." So you can't actually set `v->ops->...` in the following lines because `v->ops` should actually be a `nullptr`, right? That would be worth testing.
* Where do you delete this object again?
* My reading is that in your "clone" functions, you just let the `ops` pointer of the new object point to the `ops` pointer of the old object. if you do this, then there are potentially many objects that point to the same `ops` structure, and it becomes dangerous to call `NVFreeEmpty` on any of these `N_Vector` objects because that would free the memory of the `ops` structure but many other vectors still point to it.

I *think* that the way this is supposed to work goes like this:
```
N_Vector
create_empty_eigen_nvector()
{
  N_Vector v = N_VNewEmpty();
  assert (v->ops == nullptr);

  // Now create a global 'ops' object that will be shared among all objects for the same `VectorType` type 
  // and that all of these `N_Vector` objects point to. The 'ops' object is 'static', so it has lifetime until the end
  // of the program, and so there is no risk in letting potentially many 'N_Vectors' point to it.
  decltype(*v->ops) vec_ops;
  vec_ops.nvgetlength = NVectorOperations::N_VGetLength<VectorType>;
  [...]

  // Now let the ops pointer of the newly created vector point to it:
  v->ops = &vec_ops
}
```

You just have to make sure that in your `nvdestroy` function, you set the `ops` pointer to `nullptr` so that `N_VFreeEmpty` does not try to `delete` the object pointed to.

Reply:
In `N_VNewEmpty` the allocation does create a pointer for `v->ops`, so `assert (v->ops == nullptr)` is not right. Since `ops` is filled with functions to pointers, these are safe and I created some helper functions to assign all of the correct `ops` functions. This should also solve the ownership issue in the `clone` functions. So I did a little different implementation than you suggest but I think it works well enough.